File: packages/guards/src/pii-cn.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type PiiCnEntity =
  | 'id-card'
  | 'passport'
  | 'bank-card'
  | 'social-security'
  | 'phone';

interface PiiCnOptions {
  entities: PiiCnEntity[];
  action: 'block' | 'warn' | 'mask';
}

interface PiiCnMatch {
  type: PiiCnEntity;
  value: string;
  start: number;
  end: number;
}

const PATTERNS: Record<PiiCnEntity, RegExp> = {
  'id-card': /\b\d{17}[\dXx]\b/g,
  passport: /[EGeDSPHegdsph]\d{8}/g,
  'bank-card': /\b\d{16,19}\b/g,
  'social-security': /\b\d{18}\b/g,
  phone: /(?:1[3-9]\d{9})/g,
};

const MASK_LABELS: Record<PiiCnEntity, string> = {
  'id-card': '[身份证号]',
  passport: '[护照号]',
  'bank-card': '[银行卡号]',
  'social-security': '[社保号]',
  phone: '[手机号]',
};

const ID_WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2];
const ID_CHECK_MAP = '10X98765432';

function validateIdCard(raw: string): boolean {
  if (raw.length !== 18) return false;
  const digits = raw.toUpperCase();
  let sum = 0;
  for (let i = 0; i < 17; i++) {
    const d = Number(digits[i]);
    if (isNaN(d)) return false;
    sum += d * ID_WEIGHTS[i];
  }
  const check = ID_CHECK_MAP[sum % 11];
  return check === digits[17];
}

function isValidRegion(raw: string): boolean {
  const region = Number(raw.slice(0, 2));
  const validPrefixes = [
    11, 12, 13, 14, 15, 21, 22, 23,
    31, 32, 33, 34, 35, 36, 37, 41,
    42, 43, 44, 45, 46, 50, 51, 52,
    53, 54, 61, 62, 63, 64, 65,
  ];
  return validPrefixes.includes(region);
}

function detect(
  text: string,
  entities: PiiCnEntity[],
): PiiCnMatch[] {
  const matches: PiiCnMatch[] = [];
  for (const entity of entities) {
    const re = new RegExp(PATTERNS[entity].source, PATTERNS[entity].flags);
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      const value = m[0];
      if (entity === 'id-card') {
        if (!validateIdCard(value) || !isValidRegion(value)) continue;
      }
      matches.push({
        type: entity,
        value,
        start: m.index,
        end: m.index + value.length,
      });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskText(text: string, matches: PiiCnMatch[]): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) +
      MASK_LABELS[m.type] +
      result.slice(m.end);
  }
  return result;
}

export function piiCn(options: PiiCnOptions): Guard {
  return {
    name: 'pii-cn',
    version: '0.1.0',
    description: 'Chinese PII detection and masking',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, options.entities);
      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'pii-cn',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'pii-cn',
          passed: true,
          action: 'override',
          overrideText: maskText(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map(({ type, value }) => ({
              type,
              value,
            })),
          },
        };
      }

      return {
        guardName: 'pii-cn',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map(({ type, value }) => ({
            type,
            value,
          })),
        },
      };
    },
  };
}

File: python/open_guardrail/guards/pii_cn.py
"""Chinese PII detection and masking guard."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "id-card": re.compile(r"\b\d{17}[\dXx]\b"),
    "passport": re.compile(r"[EGeDSPHegdsph]\d{8}"),
    "bank-card": re.compile(r"\b\d{16,19}\b"),
    "social-security": re.compile(r"\b\d{18}\b"),
    "phone": re.compile(r"(?:1[3-9]\d{9})"),
}

_MASK_LABELS: dict[str, str] = {
    "id-card": "[身份证号]",
    "passport": "[护照号]",
    "bank-card": "[银行卡号]",
    "social-security": "[社保号]",
    "phone": "[手机号]",
}

_ID_WEIGHTS = [
    7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2,
]
_ID_CHECK_MAP = "10X98765432"

_VALID_REGIONS = {
    11, 12, 13, 14, 15, 21, 22, 23,
    31, 32, 33, 34, 35, 36, 37, 41,
    42, 43, 44, 45, 46, 50, 51, 52,
    53, 54, 61, 62, 63, 64, 65,
}


def _validate_id_card(raw: str) -> bool:
    if len(raw) != 18:
        return False
    upper = raw.upper()
    total = 0
    for i in range(17):
        d = upper[i]
        if not d.isdigit():
            return False
        total += int(d) * _ID_WEIGHTS[i]
    check = _ID_CHECK_MAP[total % 11]
    return check == upper[17]


def _is_valid_region(raw: str) -> bool:
    region = int(raw[:2])
    return region in _VALID_REGIONS


def _detect(
    text: str, entities: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            value = m.group()
            if entity == "id-card":
                if not _validate_id_card(value):
                    continue
                if not _is_valid_region(value):
                    continue
            matches.append(
                {
                    "type": entity,
                    "value": value,
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[PII]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _PiiCn:
    def __init__(
        self,
        *,
        action: str = "block",
        entities: Optional[List[str]] = None,
    ) -> None:
        self.name = "pii-cn"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="pii-cn",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [
            {"type": m["type"], "value": m["value"]}
            for m in matches
        ]

        if self.action == "mask":
            return GuardResult(
                guard_name="pii-cn",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        return GuardResult(
            guard_name="pii-cn",
            passed=False,
            action=self.action,
            message="Chinese PII detected in text",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains Chinese personally"
                    " identifiable information"
                ),
            },
        )


def pii_cn(
    *,
    action: str = "block",
    entities: Optional[List[str]] = None,
) -> _PiiCn:
    return _PiiCn(action=action, entities=entities)
